session.go

```diff
@@ -0,0 +1,160 @@
+package mockoidc
+
+import (
+	"strings"
+	"time"
+
+	"github.com/dgrijalva/jwt-go"
+)
+
+// TimeFunc is the default function that the token generators will use for the current time
+var TimeFunc = time.Now
+
+// TimeNowFunc is the type of the function that can be passed tot he topken generators
+// if time.Now needs to be overriden for testing purposes
+type TimeNowFunc func() time.Time
+
+// Session is a struct that maintains session state across requests
+type Session struct {
+	SessionID  string
+	Scopes     []string
+	OAuthState string
+	OIDCNonce  string
+	User       *User
+}
+
+// SessionStore is a map of #Session structs
+type SessionStore struct {
+	Store map[string]*Session
+}
+
+// Config is the OIDCMock server configuration information (here as placeholder)
+type Config struct {
+	ClientID     string
+	ClientSecret string
+	Keypair      *Keypair
+	AccessTTL    int64
+	RefreshTTL   int64
+	Issuer       string
+}
+
+// NewSession returns a pointer to a new Session struct
+func (ss *SessionStore) NewSession(scope string, oAuthState string, oidcNonce string, user *User) (*Session, error) {
+	sessionID, err := nonce(24)
+	if err != nil {
+		return nil, err
+	}
+
+	session := &Session{
+		SessionID:  sessionID,
+		Scopes:     strings.Split(scope, " "),
+		OAuthState: oAuthState,
+		OIDCNonce:  oidcNonce,
+		User:       user,
+	}
+
+	ss.Store[sessionID] = session
+
+	return session, nil
+}
+
+// AccessToken returns the JWT token with the appropriate claims for an access token
+func (s *Session) AccessToken(config *Config, timeFunc TimeNowFunc) (string, error) {
+
+	claims := standardClaims(s, config, config.AccessTTL, timeFunc)
+
+	token, err := config.Keypair.SignJWT(claims)
+	if err != nil {
+		return "", err
+	}
+
+	return token, nil
+}
+
+// standardClaims returns a populated jwt.StandardCLaims struct
+func standardClaims(s *Session, config *Config, ttl int64, timeFunc TimeNowFunc) jwt.StandardClaims {
+	if timeFunc == nil {
+		timeFunc = TimeFunc
+	}
+
+	return jwt.StandardClaims{
+		Audience:  "Audience",
+		ExpiresAt: timeFunc().UnixNano() + ttl,
+		Id:        s.SessionID,
+		IssuedAt:  timeFunc().UnixNano(),
+		Issuer:    config.Issuer,
+		NotBefore: timeFunc().UnixNano(),
+		Subject:   s.User.Email,
+	}
+}
+
+// RefreshToken returns the JWT token with the appropriate claims for an access token
+func (s *Session) RefreshToken(config *Config, timeFunc TimeNowFunc) (string, error) {
+	claims := standardClaims(s, config, config.RefreshTTL, timeFunc)
+
+	token, err := config.Keypair.SignJWT(claims)
+	if err != nil {
+		return "", err
+	}
+
+	return token, nil
+}
+
+// IDTokenClaims are the claims to be provided to jwt.Sign to create the ID token
+type IDTokenClaims struct {
```

Comment:
Here's some resources on what OneLogin has as additional claims that `profile` scope will cause to be spat out in case we want to tweak what subset we want to support: https://developers.onelogin.com/openid-connect/scopes

Reply:
I would expect we'd want to support all the scopes that OneLogin supports, unless we don't think that users are going to want to test their scopes through the mock server.  I was imagining that certain implementations might want to be able to test if certain pieces of information are returned, but maybe that's not worthwhile testing agains a mock server?

Also, is there a need to generalize beyond the OneLogin ecosystem?  I assume that is a thing?  THere is more than one OIDC provider

Reply:
Yeah, and OIDC IdPs are all over the place -- so I think we get a good sample subset.

`groups` scope for instance resulting in a `[]string` `groups` claim, that is not in the spec. And while semi-universal, enough IdPs deviated from the spec to make it a nightmare. In OAuth2 Proxy we have so much corner case code handling all the weird stuff IdPs do around groups/roles 😅